### PR TITLE
Feature/add write pipe func

### PR DESCRIPTION
**Pull request by dothebart**
### Scope & Purpose

add wrapper function to write to pipes

- [x] :pizza: New internal feature

#### Backports:

- [x] Backports required for: *(Please specify versions and link PRs)*

### Testing & Verification

- [x] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*
- [x] There are tests in an external testing repository: RTA upgrade / Hot backup tests

### CODE REVIEW

**Comment by dothebart** [comment]
write/close pipe functionality can be tested like this:
```
fs = require("fs")
const internal = require('internal');
const executeExternal = internal.executeExternal;
res = executeExternal("python3", ["-c", 'import sys; print(sys.stdin.read())'], true, [])
fs.writePipe(res.pid, "bla\n")
fs.closePipe(res.pid, false)
fs.readPipe(res.pid)
```

File: js/common/bootstrap/modules/fs.js [hunk 1]
```diff
@@ -457,14 +457,32 @@ global.DEFINE_MODULE('fs', (function () {
   }
 
   // //////////////////////////////////////////////////////////////////////////////
-  // / @brief read
+  // / @brief read from pie
```
Comment (jsteemann) [inline js/common/bootstrap/modules/fs.js:460]: :cake: 
```suggestion
  // / @brief read from pipe
```

File: lib/Basics/FileUtils.cpp [hunk 2]
```diff
@@ -739,80 +740,42 @@ void makePathAbsolute(std::string& path) {
...
+                              program, "': ",
+                              TRI_last_error());
+    THROW_ARANGO_EXCEPTION(res);
+  }
+  process = TRI_LookupSpawnedProcess(external._pid);
+  while (res = TRI_CheckExternalProcess(external, false, 0),
+         (res._status == TRI_EXT_RUNNING)) {
+    auto nRead = TRI_ReadPipe(process, buf, sizeof(buf) - 1);
+    if (nRead > 0) {
+      output += std::string(buf, nRead);
```
Comment (jsteemann) [inline lib/Basics/FileUtils.cpp:773]: Can be made more efficient without creating a temporary string:
```suggestion
      output.append(buf, nRead);
```

File: lib/Basics/FileUtils.cpp [hunk 3]
```diff
@@ -739,80 +740,42 @@ void makePathAbsolute(std::string& path) {
...
 
-    if (res != 0) {
-      throwProgramError(program);
+    LOG_TOPIC("a557b", TRACE, arangodb::Logger::FIXME)
+      << StringUtils::concatT("open failed for file '",
+                              program, "': ",
+                              TRI_last_error());
+    THROW_ARANGO_EXCEPTION(res);
+  }
+  process = TRI_LookupSpawnedProcess(external._pid);
```
Comment (jsteemann) [inline lib/Basics/FileUtils.cpp:768]: `TRI_LookupSpawnedProcess` can return a nullptr. This needs to be checked here, and reacted to.

File: lib/Basics/process-utils.cpp [hunk 4]
```diff
@@ -966,6 +966,36 @@ void TRI_CreateExternalProcess(char const* executable,
...
+      (!read && TRI_IS_INVALID_PIPE(process->_writePipe))
+      ) {
+    return;
+  }
+
+  auto pipe = (read) ? &process->_readPipe : &process->_writePipe;
+  
+#ifndef _WIN32
+  if (*pipe != -1) {
+    FILE *stream = fdopen(*pipe, "w");
```
Comment (jsteemann) [inline lib/Basics/process-utils.cpp:982]: we need to check here if `stream` is a `nullptr`!

Comment (jsteemann) [inline lib/Basics/process-utils.cpp:982]:
```suggestion
    FILE* stream = fdopen(*pipe, "w");
```

File: lib/Basics/process-utils.cpp [hunk 5]
```diff
@@ -983,6 +1013,25 @@ TRI_read_return_t TRI_ReadPipe(ExternalProcess const* process,
...
 
+////////////////////////////////////////////////////////////////////////////////
+/// @brief writes from the pipe of processes
+////////////////////////////////////////////////////////////////////////////////
+
+bool TRI_WritePipe(ExternalProcess const* process,
+                   char const* buffer,
+                   size_t bufferSize) {
+  if (process == nullptr || TRI_IS_INVALID_PIPE(process->_writePipe)) {
+    return 0;
```
Comment (jsteemann) [inline lib/Basics/process-utils.cpp:1024]:
```suggestion
    return false;
```

File: lib/Basics/process-utils.cpp [hunk 6]
```diff
@@ -966,6 +966,36 @@ void TRI_CreateExternalProcess(char const* executable,
...
+    return;
+  }
+
+  auto pipe = (read) ? &process->_readPipe : &process->_writePipe;
+  
+#ifndef _WIN32
+  if (*pipe != -1) {
+    FILE *stream = fdopen(*pipe, "w");
+    fflush(stream);
+    close(*pipe);
```
Comment (jsteemann) [inline lib/Basics/process-utils.cpp:984]: In theory, `close(fd)` can fail, and we are not checking for it.
However, in reality there is not so much that can be done if it fails, except to log an error or such.

File: Installation/Windows/Templates/NSIS.template.in [hunk 7]
```diff
@@ -650,6 +650,8 @@ FunctionEnd
 
 Function InstallOptionsPage2
    IfSilent 0 continueUI
+   ; this would be called inside the UI code:
+   Call stop_old_service
```
Comment (jsteemann) [inline Installation/Windows/Templates/NSIS.template.in:654]: I have no idea if this change is related to the other changes in this PR, and if it is required.

File: lib/V8/v8-utils.cpp [hunk 8]
```diff
@@ -3036,6 +3036,79 @@ static void JS_ReadPipe(v8::FunctionCallbackInfo<v8::Value> const& args) {
...
+        "not allowed to execute or modify state of external processes");
+  }
+
+  TRI_pid_t pid = static_cast<TRI_pid_t>(TRI_ObjectToInt64(isolate, args[0]));
+  ExternalProcess const* proc = TRI_LookupSpawnedProcess(pid);
+
+  if (proc == nullptr) {
+    TRI_V8_THROW_EXCEPTION_MESSAGE(
+      TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND,
+      "didn't find the process identified by <external-identifier> to read the pipe from.");
```
Comment (jsteemann) [inline lib/V8/v8-utils.cpp:3066]:
```suggestion
      "didn't find the process identified by <external-identifier> to write the pipe into.");
```

File: lib/V8/v8-utils.cpp [hunk 9]
```diff
@@ -3036,6 +3036,79 @@ static void JS_ReadPipe(v8::FunctionCallbackInfo<v8::Value> const& args) {
...
+    TRI_V8_RETURN_TRUE();
+  }
+  TRI_V8_RETURN_FALSE();
+  TRI_V8_TRY_CATCH_END
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief was docuBlock JS_Read
+////////////////////////////////////////////////////////////////////////////////
+
```
Comment (jsteemann) [inline lib/V8/v8-utils.cpp:3080]:
```suggestion
```

File: lib/V8/v8-utils.cpp [hunk 10]
```diff
@@ -3036,6 +3036,79 @@ static void JS_ReadPipe(v8::FunctionCallbackInfo<v8::Value> const& args) {
...
+static void JS_ClosePipe(v8::FunctionCallbackInfo<v8::Value> const& args) {
+  TRI_V8_TRY_CATCH_BEGIN(isolate)
+  v8::HandleScope scope(isolate);
+
+  if (args.Length() != 2) {
+    TRI_V8_THROW_EXCEPTION_USAGE("writePipe(<external-identifier>, bool Read)");
```
Comment (jsteemann) [inline lib/V8/v8-utils.cpp:3086]:
```suggestion
    TRI_V8_THROW_EXCEPTION_USAGE("closePipe(<external-identifier>, bool read)");
```

File: lib/V8/v8-utils.cpp [hunk 11]
```diff
@@ -3036,6 +3036,79 @@ static void JS_ReadPipe(v8::FunctionCallbackInfo<v8::Value> const& args) {
...
+        "not allowed to execute or modify state of external processes");
+  }
+
+  TRI_pid_t pid = static_cast<TRI_pid_t>(TRI_ObjectToInt64(isolate, args[0]));
+  ExternalProcess* proc = TRI_LookupSpawnedProcess(pid);
+
+  if (proc == nullptr) {
+    TRI_V8_THROW_EXCEPTION_MESSAGE(
+      TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND,
+      "didn't find the process identified by <external-identifier> to read the pipe from.");
```
Comment (jsteemann) [inline lib/V8/v8-utils.cpp:3104]:
```suggestion
      "didn't find the process identified by <external-identifier> to close the pipe for.");
```

File: lib/V8/v8-utils.cpp [hunk 12]
```diff
@@ -3036,6 +3036,79 @@ static void JS_ReadPipe(v8::FunctionCallbackInfo<v8::Value> const& args) {
...
+  }
+
+  bool read = TRI_ObjectToBoolean(isolate, args[1]);
+  TRI_ClosePipe(proc, read);
+  TRI_V8_TRY_CATCH_END
```
Comment (jsteemann) [inline lib/V8/v8-utils.cpp:3105]: Does this JS function have any return value? Should it?

Comment (dothebart) [reply lib/V8/v8-utils.cpp:3105]: TRI_ClosePipe is a void function? hence, if there are errors on flushing / closing we ignore it I guess?

File: Installation/Windows/Templates/NSIS.template.in (hunk 7 above)
Comment (jsteemann) [reply Installation/Windows/Templates/NSIS.template.in:654]: Should it be removed then?

File: lib/Basics/FileUtils.cpp (hunk 3 above)
Comment (jsteemann) [reply lib/Basics/FileUtils.cpp:768]: nullptr is not handled here...

**Review by jsteemann — APPROVED** [review]
LGTM!

**Pull request merged**